1990 New Mexico Congressional Districts

jerrydai08 opened the pull request:
# 1990 New Mexico Congressional Districts

## Redistricting requirements
In New Mexico, we consult [A GUIDE TO STATE AND CONGRESSIONAL REDISTRICTING IN NEW MEXICO](https://www.nmlegis.gov/Redistricting/Documents/134250.pdf) and impose the following constraints. In our simulations, districts must:

1. be contiguous
1. have equal populations
1. be geographically compact
1. preserve county and municipality boundaries as much as possible
1. satisfy the Voting Rights Act


### Algorithmic Constraints
We enforce a maximum population deviation of 0.5%.

## Data Sources
Data for New Mexico comes from the [ALARM Project's update](https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/ZV5KF3) to [The Record of American Democracy](https://road.hmdc.harvard.edu/).

## Pre-processing Notes
No manual pre-processing decisions were necessary.

## Simulation Notes
We sample 10,000 districting plans for New Mexico across 5 independent runs of the SMC algorithm.
We then thinned the number of samples to 5,000. 
No special techniques were needed to produce the sample.

## Validation
<img width="3000" height="5400" alt="validation_20251212_1425" src="https://github.com/user-attachments/assets/cca5c8a6-8c88-41ed-8c6b-62d6988d2cb0" />

```
SMC: 10,000 sampled plans of 3 districts on 390 units
`adapt_k_thresh`=0.99 • `seq_alpha`=0.9
`pop_temper`=0.05

Plan diversity 80% range: 0.16 to 0.68
✖ WARNING: Low plan diversity

R-hat values for summary statistics:
        pop_overlap           total_vap            plan_dev           comp_edge         comp_polsby     comp_bbox_reock 
              1.001               1.002               1.002               1.000               1.000               1.002 
          pop_other           pop_white           pop_black            pop_hisp            pop_aian           pop_asian 
              1.001               1.001               1.001               1.000               1.001               1.002 
           vap_hisp            vap_aian           vap_other           vap_white           vap_black           vap_asian 
              1.001               1.001               1.001               1.001               1.001               1.001 
      county_splits total_county_splits         muni_splits   total_muni_splits                 ndv                 nrv 
              1.001               1.001               1.001               1.001               1.001               1.002 
            ndshare 
              1.002 

Sampling diagnostics for SMC run 1 of 10 (2,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k 
Split 1     1,849 (92.4%)      5.2%        0.35 1,259 (100%)      7 
Split 2     1,511 (75.6%)      3.0%        0.51 1,155 ( 91%)      5 
Resample    1,382 (69.1%)       NA%        0.50 1,603 (127%)     NA 

Sampling diagnostics for SMC run 2 of 10 (2,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k 
Split 1     1,846 (92.3%)      5.2%        0.35 1,286 (102%)      7 
Split 2     1,498 (74.9%)      3.0%        0.50 1,139 ( 90%)      5 
Resample    1,446 (72.3%)       NA%        0.50 1,636 (129%)     NA 

Sampling diagnostics for SMC run 3 of 10 (2,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k 
Split 1     1,854 (92.7%)      7.7%        0.35 1,258 (100%)      5 
Split 2     1,513 (75.6%)      3.8%        0.50 1,165 ( 92%)      4 
Resample    1,465 (73.3%)       NA%        0.49 1,625 (129%)     NA 

Sampling diagnostics for SMC run 4 of 10 (2,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k 
Split 1     1,842 (92.1%)      5.3%        0.36 1,251 ( 99%)      7 
Split 2     1,309 (65.4%)      3.6%        0.51 1,164 ( 92%)      4 
Resample    1,109 (55.5%)       NA%        0.51 1,612 (128%)     NA 

Sampling diagnostics for SMC run 5 of 10 (2,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k 
Split 1     1,843 (92.1%)      5.3%        0.36 1,263 (100%)      7 
Split 2     1,648 (82.4%)      3.6%        0.48 1,161 ( 92%)      4 
Resample    1,596 (79.8%)       NA%        0.47 1,640 (130%)     NA 

Sampling diagnostics for SMC run 6 of 10 (2,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k 
Split 1     1,845 (92.2%)      7.6%        0.36 1,250 ( 99%)      5 
Split 2     1,435 (71.7%)      4.9%        0.52 1,145 ( 91%)      3 
Resample    1,375 (68.8%)       NA%        0.51 1,608 (127%)     NA 

Sampling diagnostics for SMC run 7 of 10 (2,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k 
Split 1     1,847 (92.4%)      6.3%        0.35 1,266 (100%)      6 
Split 2     1,240 (62.0%)      3.6%        0.52 1,158 ( 92%)      4 
Resample      957 (47.8%)       NA%        0.52 1,568 (124%)     NA 

Sampling diagnostics for SMC run 8 of 10 (2,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k 
Split 1     1,841 (92.1%)      7.5%        0.36 1,241 ( 98%)      5 
Split 2     1,625 (81.3%)      4.7%        0.49 1,173 ( 93%)      3 
Resample    1,544 (77.2%)       NA%        0.49 1,649 (130%)     NA 

Sampling diagnostics for SMC run 9 of 10 (2,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k 
Split 1     1,840 (92.0%)      5.2%        0.36 1,237 ( 98%)      7 
Split 2     1,523 (76.2%)      3.7%        0.52 1,176 ( 93%)      4 
Resample    1,405 (70.2%)       NA%        0.51 1,592 (126%)     NA 

Sampling diagnostics for SMC run 10 of 10 (2,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k 
Split 1     1,848 (92.4%)      5.2%        0.36 1,248 ( 99%)      7 
Split 2     1,623 (81.1%)      3.4%        0.50 1,154 ( 91%)      4 
Resample    1,554 (77.7%)       NA%        0.50 1,598 (126%)     NA 

•  Watch out for low effective samples, very low acceptance rates (less than 1%), large std. devs. of the log weights (more than 3 or
so), and low numbers of unique plans. R-hat values for summary statistics should be between 1 and 1.05.
• Low diversity: Check for potential bottlenecks. Increase the number of samples. Examine the diversity plot with
`hist(plans_diversity(plans), breaks=24)`. Consider weakening or removing constraints, or increasing the population tolerance. If the
acceptance rate drops quickly in the final splits, try increasing `pop_temper` by 0.01.
```

## Checklist

- [X] I have followed the [instructions](https://github.com/alarm-redist/fifty-states/blob/main/CONTRIBUTING.md)
- [X] I have updated the [tracker](https://docs.google.com/spreadsheets/d/1k_tYLoE49W_DCK1tcWbouoYZFI9WD76oayEt5TOmJg4/edit#gid=453387933)
- [X] All `TODO` lines from the template code have been removed
- [X] I have merged in the main branch and then recalculated summary statistics
- [X] I have run `enforce_style()` to format my code
- [x] The documentation copied above is up-to-date 
- [X] There are no data files in this pull request
- [X] None of the file output paths (for the `redist_map` and `redist_plans` objects, and summary statistics) have been edited

@tylersimko

jerrydai08 committed:
1990 New Mexico Congressional Districts

jerrydai08 committed:
1990 New Mexico Congressional Districts

jerrydai08 committed:
Update

kosukeimai commented:
@jerrydai08 Can you please specify the reviewer?

kosukeimai commented:
Good. @tylersimko Please take a look.

tylersimko commented:
Thanks @jerrydai08 . It looks like those redistricting instructions are from 2000, right? Not a problem if we can't find public 1990 ones, but we should mention that in documentation. A few thoughts:

1. Agree diversity is lower than ideal, but this this looks in line with other NM years like [2010](https://github.com/alarm-redist/fifty-states/pull/148). It's a tricky state, probably due to such disparity in population density across the state. Looks good to me.
2. Have you confirmed performance of that MMD is similar to enacted plan?

jerrydai08 commented:
Yes, the redistricting instructions are from 2000: I couldn't find 1990 ones, so I defaulted to 2000 instructions. I'll update that in the documentation.

By the performance of the MMD, do you mean how the MMD votes? I saw that the MVAP ratios were fairly similar to the enacted plan (which is quite rare, usually it's a lot harder to get MMDs because it seems like the districting around those is so intentional), and that was the main metric I used to judge the MMD.

jerrydai08 committed:
Todo update

jerrydai08 committed:
Update 01_prep_NM_cd_1990.R

jerrydai08 commented:
Redid the projection, updated validation:

<img width="3000" height="5400" alt="validation_20260221_1216" src="https://github.com/user-attachments/assets/c027ff40-3451-4d41-b7ac-2b923678ecb3" />

jerrydai08 commented:
Here are the updated summary statistics:

```
SMC: 10,000 sampled plans of 3 districts on 390 units
`adapt_k_thresh`=0.99 • `seq_alpha`=0.9
`pop_temper`=0.05

Plan diversity 80% range: 0.16 to 0.62
✖ WARNING: Low plan diversity

R-hat values for summary statistics:
        pop_overlap           total_vap            plan_dev           comp_edge         comp_polsby     comp_bbox_reock 
              1.001               1.002               1.002               1.000               1.000               1.002 
          pop_other           pop_white           pop_black            pop_hisp            pop_aian           pop_asian 
              1.001               1.001               1.001               1.000               1.001               1.002 
           vap_hisp            vap_aian           vap_other           vap_white           vap_black           vap_asian 
              1.001               1.001               1.001               1.001               1.001               1.001 
      county_splits total_county_splits         muni_splits   total_muni_splits                 ndv                 nrv 
              1.001               1.001               1.001               1.001               1.001               1.002 
            ndshare 
              1.002 

Sampling diagnostics for SMC run 1 of 10 (2,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k 
Split 1     1,849 (92.4%)      5.2%        0.35 1,259 (100%)      7 
Split 2     1,511 (75.6%)      3.0%        0.51 1,155 ( 91%)      5 
Resample    1,382 (69.1%)       NA%        0.50 1,603 (127%)     NA 

Sampling diagnostics for SMC run 2 of 10 (2,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k 
Split 1     1,846 (92.3%)      5.2%        0.35 1,286 (102%)      7 
Split 2     1,498 (74.9%)      3.0%        0.50 1,139 ( 90%)      5 
Resample    1,446 (72.3%)       NA%        0.50 1,636 (129%)     NA 

Sampling diagnostics for SMC run 3 of 10 (2,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k 
Split 1     1,854 (92.7%)      7.7%        0.35 1,258 (100%)      5 
Split 2     1,513 (75.6%)      3.8%        0.50 1,165 ( 92%)      4 
Resample    1,465 (73.3%)       NA%        0.49 1,625 (129%)     NA 

Sampling diagnostics for SMC run 4 of 10 (2,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k 
Split 1     1,842 (92.1%)      5.3%        0.36 1,251 ( 99%)      7 
Split 2     1,309 (65.4%)      3.6%        0.51 1,164 ( 92%)      4 
Resample    1,109 (55.5%)       NA%        0.51 1,612 (128%)     NA 

Sampling diagnostics for SMC run 5 of 10 (2,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k 
Split 1     1,843 (92.1%)      5.3%        0.36 1,263 (100%)      7 
Split 2     1,648 (82.4%)      3.6%        0.48 1,161 ( 92%)      4 
Resample    1,596 (79.8%)       NA%        0.47 1,640 (130%)     NA 

Sampling diagnostics for SMC run 6 of 10 (2,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k 
Split 1     1,845 (92.2%)      7.6%        0.36 1,250 ( 99%)      5 
Split 2     1,435 (71.7%)      4.9%        0.52 1,145 ( 91%)      3 
Resample    1,375 (68.8%)       NA%        0.51 1,608 (127%)     NA 

Sampling diagnostics for SMC run 7 of 10 (2,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k 
Split 1     1,847 (92.4%)      6.3%        0.35 1,266 (100%)      6 
Split 2     1,240 (62.0%)      3.6%        0.52 1,158 ( 92%)      4 
Resample      957 (47.8%)       NA%        0.52 1,568 (124%)     NA 

Sampling diagnostics for SMC run 8 of 10 (2,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k 
Split 1     1,841 (92.1%)      7.5%        0.36 1,241 ( 98%)      5 
Split 2     1,625 (81.3%)      4.7%        0.49 1,173 ( 93%)      3 
Resample    1,544 (77.2%)       NA%        0.49 1,649 (130%)     NA 

Sampling diagnostics for SMC run 9 of 10 (2,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k 
Split 1     1,840 (92.0%)      5.2%        0.36 1,237 ( 98%)      7 
Split 2     1,523 (76.2%)      3.7%        0.52 1,176 ( 93%)      4 
Resample    1,405 (70.2%)       NA%        0.51 1,592 (126%)     NA 

Sampling diagnostics for SMC run 10 of 10 (2,000 samples)
         Eff. samples (%) Acc. rate Log wgt. sd  Max. unique Est. k 
Split 1     1,848 (92.4%)      5.2%        0.36 1,248 ( 99%)      7 
Split 2     1,623 (81.1%)      3.4%        0.50 1,154 ( 91%)      4 
Resample    1,554 (77.7%)       NA%        0.50 1,598 (126%)     NA 

•  Watch out for low effective samples, very low acceptance rates (less than 1%), large std. devs. of the log weights (more than 3 or
so), and low numbers of unique plans. R-hat values for summary statistics should be between 1 and 1.05.
• Low diversity: Check for potential bottlenecks. Increase the number of samples. Examine the diversity plot with
`hist(plans_diversity(plans), breaks=24)`. Consider weakening or removing constraints, or increasing the population tolerance. If the
acceptance rate drops quickly in the final splits, try increasing `pop_temper` by 0.01.
```

tylersimko commented:
Looks good to me, thanks @jerrydai08 !

tylersimko commented:
@jerrydai08 When you upload, please make sure to edit so that there are 5k plans (not 10k).
